### PR TITLE
feature(nx-plugin): adds playwright executor and generators

### PR DESCRIPTION
**Pull request by bsunderhus**
Adds `playwright` for component test according to [documentation](https://playwright.dev/docs/test-components)

1. executor:
    * `playwright` (if `testingType === component` then execute task manually, otherwise delegate to `@nx/playwright`)
2. generators:
    * `playwright-component-configuration` (as per https://nx.dev/nx-api/react/generators/cypress-component-configuration)
    * `component-playwright-spec` (as per https://nx.dev/nx-api/react/generators/component-cypress-spec)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -67,6 +67,9 @@ yarn nx generate library <name>
 # Configures storybook for the package
 yarn nx generate configure-storybook <name>
 
+# Configures playwright for the package
```
Comment (ling1726) [inline CONTRIBUTING.md:70]:
```suggestion
# Configures playwright browser testing for the package
```

File: packages/nx-plugin/src/generators/configure-playwright/files/src/index.spec.tsx.template [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
```
Comment (Hotell) [inline packages/nx-plugin/src/generators/configure-playwright/files/src/index.spec.tsx.template:1]: this should be part of component separate generator, similar to https://nx.dev/nx-api/react/generators/component-cypress-spec but for playwright ofc.

can you create feature request on nx repo to add component testing capabilities to core nx playwright plugin ?

File: packages/nx-plugin/generators.json [hunk 3]
```diff
@@ -10,6 +10,11 @@
       "schema": "./src/generators/configure-storybook/schema.json",
       "description": "configure-storybook generator"
     },
+    "configure-playwright": {
```
Comment (Hotell) [inline packages/nx-plugin/generators.json:13]: I'd recommend follow existing patterns to be explicit what it does 

following https://nx.dev/nx-api/react/generators/cypress-component-configuration -> `playwright-component-configuration`

File: packages/nx-plugin/src/executors/playwright-test/executor.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { execSync } from 'child_process';
+import { ExecutorContext } from '@nx/devkit';
+import { PlaywrightTestExecutorSchema } from './schema';
+
+export default async function runExecutor(
```
Comment (Hotell) [inline packages/nx-plugin/src/executors/playwright-test/executor.ts:5]: can we tweak the api to follow something like this ? ( we can omit invoking nx core executor in order to keep this simple for now )

image

File: packages/nx-plugin/src/generators/configure-playwright/files/src/index.spec.tsx.template (hunk 2 above)
Comment (bsunderhus) [reply packages/nx-plugin/src/generators/configure-playwright/files/src/index.spec.tsx.template:1]: Here's the feature request for nx https://github.com/nrwl/nx/issues/21594

Comment (bsunderhus) [reply packages/nx-plugin/src/generators/configure-playwright/files/src/index.spec.tsx.template]: Ok, now we got 2 generators and 1 executor:

1. executor:
    * `nx-plugin:playwright`
2. generators:
    * `playwright-component-configuration` (as per https://nx.dev/nx-api/react/generators/cypress-component-configuration)
    * `component-playwright-spec` (as per https://nx.dev/nx-api/react/generators/component-cypress-spec)

File: packages/nx-plugin/src/executors/playwright/schema.d.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { PlaywrightExecutorSchema as NxPlaywrightExecutorSchema } from '@nx/playwright';
+
+export interface PlaywrightExecutorSchema extends NxPlaywrightExecutorSchema {
+  testingType?: 'component';
+} // eslint-disable-line
```
Comment (Hotell) [inline packages/nx-plugin/src/executors/playwright/schema.d.ts:5]: typo ? 

```suggestion
}
```

File: packages/nx-plugin/src/generators/playwright-component-configuration/generator.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  targets['component-test'] = {
+    executor: '@fluentui-contrib/nx-plugin:playwright',
+    options: {
+      testingType: 'component',
+    },
+  };
+
+  updateProjectConfiguration(tree, name, project);
+
+  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, options);
```
Comment (Hotell) [inline packages/nx-plugin/src/generators/playwright-component-configuration/generator.ts:36]: there are more explicit APIs to update project targets
 
```suggestion
  const config = readProjectConfiguration(tree, name);

  const targetDefinition = {
    executor: '@fluentui-contrib/nx-plugin:playwright',
    options: {
      testingType: 'component',
    },
  };

  config.targets = config.targets ?? {};
  config.targets['component-test'] = targetDefinition;

  updateProjectConfiguration(tree, name, { ...config });

  generateFiles(tree, path.join(__dirname, 'files'), config.root, options);
```

File: packages/nx-plugin/src/executors/playwright/schema.d.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { PlaywrightExecutorSchema as NxPlaywrightExecutorSchema } from '@nx/playwright';
+
+export interface PlaywrightExecutorSchema extends NxPlaywrightExecutorSchema {
+  testingType?: 'component';
```
Comment (Hotell) [inline packages/nx-plugin/src/executors/playwright/schema.d.ts:4]: following cypress executor official plugin APIs 

```suggestion
/**
   * @default 'component'
   */
  testingType?: 'component' | 'e2e';
```

File: packages/nx-plugin/src/executors/playwright/executor.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { execSync } from 'child_process';
+import { ExecutorContext } from '@nx/devkit';
+import { PlaywrightExecutorSchema } from './schema';
+import { playwrightExecutor } from '@nx/playwright';
+
+export default async function runExecutor(
+  options: PlaywrightExecutorSchema,
+  context: ExecutorContext
+) {
+  if (options.testingType !== 'component') {
```
Comment (Hotell) [inline packages/nx-plugin/src/executors/playwright/executor.ts:10]: based on schema adjustments 

```suggestion
if (options.testingType === 'e2e') {    
```

File: packages/nx-plugin/src/executors/playwright/schema.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "http://json-schema.org/schema",
+  "version": 2,
+  "title": "Playwright component test executor",
+  "description": "",
+  "type": "object",
+  "properties": {},
```
Comment (Hotell) [inline packages/nx-plugin/src/executors/playwright/schema.json:7]: let's make better job for `--help` flag

image


```suggestion
"properties": {
    "testingType": {
      "type": "string",
      "description": "Specify the type of tests to execute.",
      "enum": ["component", "e2e"],
      "default": "component"
    }
  },
```

File: packages/nx-plugin/src/generators/component-playwright-spec/generator.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  const { name } = options;
+  const projects = getProjects(tree);
+  const project = projects.get(name);
+  if (!project) {
+    return;
+  }
+
+  const { root: projectRoot } = project;
+
+  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, options);
```
Comment (Hotell) [inline packages/nx-plugin/src/generators/component-playwright-spec/generator.ts:18]: we can simplify again

```suggestion
  const { name } = options;
  const projectConfig = readProjectConfiguration(tree, name);

  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    projectConfig.root,
    options
  );
```

File: packages/nx-plugin/src/generators/component-playwright-spec/files/src/components/__componentName__/__componentName__.spec.tsx.template [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import * as React from 'react';
+import { test, expect } from '@playwright/experimental-ct-react';
+
+test.use({ viewport: { width: 500, height: 500 } });
+
+test('should work', async ({ mount }) => {
+  const component = await mount(<div>Learn React</div>);
```
Comment (Hotell) [inline packages/nx-plugin/src/generators/component-playwright-spec/files/src/components/__componentName__/__componentName__.spec.tsx.template:7]: not sure if this is the best default ? maybe rendering the actual control would make more sense with the test failing. I'd rather avoid situations that folks generate tests that have no value but are passing.

**Review by Hotell — COMMENTED** [review]
awesome improvements ! added some final touches and lets ship it!

ty

File: packages/nx-plugin/src/executors/playwright/executor.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  if (!context.projectsConfigurations || !context.projectName) {
+    throw new Error('no project configurations');
+  }
+
+  const { root } = context.projectsConfigurations.projects[context.projectName];
+
+  if (!root) {
+    return;
```
[the remark this answers is not in the document]
Comment (bsunderhus) [reply packages/nx-plugin/src/executors/playwright/executor.ts:21]: Probably, I'm not that experienced with `nx` so I'm relying on the patterns I'm seeing on the other files:

https://github.com/microsoft/fluentui-contrib/blob/3de28ae150fb8a004eafa09f20e62740f8d84dde/packages/nx-plugin/src/executors/build/executor.ts#L26-L28

File: packages/nx-plugin/src/generators/playwright-component-configuration/generator.ts (hunk 6 above)
Comment (bsunderhus) [reply packages/nx-plugin/src/generators/playwright-component-configuration/generator.ts:36]: Thx, I'll update it accordingly. I was following the pattern I've seen in the storybook configuration, might be nice to update there also in a separate PR

https://github.com/microsoft/fluentui-contrib/blob/3de28ae150fb8a004eafa09f20e62740f8d84dde/packages/nx-plugin/src/generators/configure-storybook/generator.ts#L10-L26

**Review by Hotell — APPROVED** [review]
THANK YOU 🙏

**Pull request merged**